search/utm_mgrs_coords_match.cpp

```diff
@@ -0,0 +1,250 @@
+            break;
+    }
+
+    return pos;
+}
+
+bool IsSpace(char ch)
+{
+    return ch == ' ' || ch == '\t' || ch == '\r';
+}
+
+// Parse MGRS format "(\d\d\W)\s*(\W\W)\s*(\d+)\s*(\d+)" and converts it to lat,lon.
+// Returns true if parsed successfully or false otherwise.
+// See utm_mgrs_coords_match_test.cpp for sample MGRS strings
+bool MatchMGRSCoords(std::string const & query, double & lat, double & lon)
+{
+    long zone_code;
+    char zone_letter;
+    char square_code[2];
+    string eastingStr;
+    string northingStr;
+    long easting;
+    long northing;
+
+    strings::SimpleTokenizer it(query, " \t\r");
+    if (!it)
+        return false;
+
+    auto token = std::string(*it);
+    // Parse 2 digit zone code and 1 char zone letter
+    if (token.size() >= 3)
+    {
+        char dig1 = token[0];
+        char dig2 = token[1];
+        if (dig1 < '0' || dig1 > '9' || dig2 < '0' || dig2 > '9')
+            return false;
+
+        zone_code = (dig1 - '0') * 10 + (dig2 - '0');
+        if (zone_code<1 || zone_code > 60)
+            return false;
+
+        zone_letter = token[2];
+        token = token.substr(3);
+    }
+    else
+        return false;
+
+    // Read next token if needed.
+    if (token.size() == 0)
+    {
+        ++it;
+        if (!it)
+            return false;
+        token = std::string(*it);
+    }
+
+    // Parse 2 chars zone code.
+    if (token.size() >= 2)
+    {
+        square_code[0] = token[0];
+        square_code[1] = token[1];
+        token = token.substr(2);
+    }
+
+    // Read next token if needed.
+    if (token.size() == 0)
+    {
+        ++it;
+        if (!it)
+            return false;
+        token = std::string(*it);
+    }
+
+    // Parse easting and norhing.
+    eastingStr = token;
+
+    // Get next token if available.
+    ++it;
+    if (it)
+        northingStr = std::string(*it);
+
+    // Convert eastingStr & northingStr to numbers.
+    if (northingStr.empty())
+    {
+        // eastingStr contains both easting and northing. Let's split
+        if (eastingStr.size()%2 != 0)
+            return false;
+
+        int eastingSize = eastingStr.size()/2;
+        northingStr = eastingStr.substr(eastingSize);
+        eastingStr = eastingStr.substr(0, eastingSize);
+    }
+
+    if (eastingStr.size() != northingStr.size() || eastingStr.size()>5 || northingStr.size()>5)
+        return false;
+
+    if (!parseLong(eastingStr, easting))
+        return false;
+    if (eastingStr.size() < 5)
+    {
+        int decShift = 5 - eastingStr.size();
+        easting *= pow(10L, decShift);
+    }
+
+    if (!parseLong(northingStr, northing))
+        return false;
+    if (northingStr.size() < 5)
+    {
+        int decShift = 5 - northingStr.size();
+        northing *= pow(10L, decShift);
+    }
+
+    return utm_mgrs_utils::MGRStoLatLon((double)easting, (double)northing, zone_code, zone_letter, square_code, lat, lon);
+}
+
+bool parseLong(string const & txt, long & value)
```

Comment:
String to long conversion. Am I reinventing the wheel here?

Reply:
strings::to_int64